async-book/src/getting_started/http_server_example.md

```diff
@@ -0,0 +1,199 @@
+        // implementation of the the Hyper `Service` trait, which is an
+        // asynchronous function from a generic `Request` to a `Response`.
+        service::service_fn,
+
+        // A function which runs a future to completion using the Hyper runtime.
+        rt::run,
+    },
+    futures::{
+        // `TokioDefaultSpawner` tells futures 0.3 futures how to spawn tasks
+        // onto the Tokio runtime.
+        compat::TokioDefaultSpawner,
+
+        // Extension traits providing additional methods on futures.
+        // `FutureExt` adds methods that work for all futures, whereas
+        // `TryFutureExt` adds methods to futures that return `Result` types.
+        future::{FutureExt, TryFutureExt},
+    },
+    std::net::SocketAddr,
+
+    // This is the redefinition of the await! macro which supports both
+    // futures 0.1 (used by Hyper and Tokio) and futures 0.3 (the new API
+    // exposed by `std::future` and implemented by `async fn` syntax).
+    tokio::await,
+};
+```
+
+Once the imports are out of the way, we can start putting together the
+boilerplate to allow us to serve requests:
+
+```rust
+async fn serve_req(req: Request<Body>) -> Result<Response<Body>, hyper::Error> {
+    unimplemented!()
+}
+
+async fn run_server(addr: SocketAddr) {
+    println!("Listening on http://{}", addr);
+
+    // Create a server bound on the provided address
+    let serve_future = Server::bind(&addr)
+        // Serve requests using our `async serve_req` function.
+        // `serve` takes a closure which returns a type implementing the
+        // `Service` trait. `service_fn` returns a value implementing the
+        // `Service` trait, and accepts a closure which goes from request
+        // to a future of the response. In order to use our `serve_req`
+        // function with Hyper, we have to box it and put it in a compatability
+        // wrapper to go from a futures 0.3 future (the kind returned by
+        // `async fn`) to a futures 0.1 future (the kind used by Hyper).
+        .serve(|| service_fn(|req|
+            serve_req(req).boxed().compat(TokioDefaultSpawner)
+        ));
+
+    // Wait for the server to complete serving or exit with an error.
+    // If an error occurred, print it to stderr.
+    if let Err(e) = await!(serve_future) {
+        eprintln!("server error: {}", e);
+    }
+}
+
+fn main() {
+    // Set the address to run our socket on.
+    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+
+    // Call our run_server function, which returns a future.
+    // As with every `async fn`, we need to run that future in order for
+    // `run_server` to do anything. Additionally, since `run_server` is an
+    // `async fn`, we need to convert it from a futures 0.3 future into a
+    // futures 0.1 future.
+    let futures_03_future = run_server(addr);
+    let futures_01_future =
+        futures_03_future.unit_error().boxed().compat(TokioDefaultSpawner);
```

Comment:
I'm guessing that once futures 0.3 becomes stable these examples will be updated. Is that being tracked somewhere? I imagine going through the entire book would mean that some might be missed.

Reply:
These particular examples aren't marked anywhere. Yes, we'll have to go through everything and make appropriate changes. I'm not sure how best to track that. Perhaps we could keep a "golden" compile-able version of each of the examples in this repo, so that as things change and we have to update the compile-able version, we know to update the book version as well. Alternatively, we could set up something like the official Rust book has that will extract all of the code examples directly from the markdown and actually compile them.